### PR TITLE
Resolving block references in prefect.yaml and adding keypath feature for blocks

### LINKED ISSUE #10388: Block and variable references are not resolved in `prefect.yaml` outside of deploy steps
### First check

- [X] I added a descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the Prefect documentation for this issue.
- [X] I checked that this issue is related to Prefect and not one of its dependencies.

### Bug summary

When blocks and/or variables are used to populate a field in `prefect.yaml` outside of `build`, `push`, or `pull`, the block or variable used is not resolved and the expected value is not inserted.

### Reproduction

Example `prefect.yaml`:

```yaml
prefect-version: 2.10.20
name: analyst-operations-qep-yaml # ex: position-pnl-pipeline-yaml

pull:
    - prefect.deployments.steps.git_clone:
        repository: https://github.com/***/***-***.git
        branch: "main"
        credentials: "{{ prefect.blocks.github-credentials.sra-service-account }}"

definitions:
    work_pools:
        ecs_work_pool: &ecs_work_pool
            name: "{{ prefect.blocks.string.ecs-work-pool }}"
            work_queue_name: default
            job_variables:
                image: "*****.dkr.ecr.us-east-2.amazonaws.com/py310-base:0.0.3" 

        ny4_work_pool: &ny4_work_pool
            name: "{{ prefect.blocks.string.ny4-work-pool }}"
            work_queue_name: default
            job_variables:
                image: "****.dkr.ecr.us-east-2.amazonaws.com/py310-base:0.0.3" 

deployments:
    - name: Set QEP Status
      description: null # optional
      entrypoint: QEP/setQEPstatus.py:set_qep_status 
      work_pool: *ecs_work_pool
      parameters: {} # optional parameters overrride here
      schedule:
        cron: "1 19 * * *"
        timezone: "America/Chicago"
```

### Error

From CLI:

This deployment configuration references work pool '{{ prefect.blocks.string.ecs-work-pool }}' which does not exist. This means no worker will be able to pick up its runs. You can create a work pool in the Prefect UI.

### Versions

```Text
Version:             2.11.3
API version:         0.8.4
Python version:      3.10.6
Git commit:          3a400865
Built:               Thu, Aug 3, 2023 3:24 PM
OS/Arch:             darwin/x86_64
Profile:             default
Server type:         cloud
```

### Additional context

_No response_

### LINKED ISSUE #10930: Resolve Keys in Block References in prefect.yaml
### First check

- [X] I added a descriptive title to this issue.
- [X] I used the GitHub search to find a similar request and didn't find it.
- [X] I searched the Prefect documentation for this feature.

### Prefect Version

2.x

### Describe the current behavior

I'm currently solving #10388. While doing so, I noticed that I can only insert the full content of the `value` field of a block.

### Describe the proposed behavior

It might be nice to be able to reference a keypath in Blocks which store JSON-like or as dict parseable content.

### Example Use

Lets say I have a JSON Block named `json-block` specifying some deployment info with the following content:

```json
{
  "name": "nice-work-pool",
  "work-queues": ["wq1", "wq2"],
  "job-vars": {
     "image": "cool-image"
   }
}
```

I think it would be cool to be able to reference the keys and get their values in `prefect.yaml`. One could use `value` as a special key to indicate a keypath. `prefect.utilities.collections.get_from_dict` also made me realize that there is already functionality to get list-values as well.

```yaml
# prefect.yaml

...
deployments:
   - name: first-deployment
     ...
     work_pool:
       name: "{{ prefect.blocks.json.json-block.value.name }}"  # nice-work-pool
       work_queue_name: "{{ prefect.blocks.json-block.value.work-queues[1]"  # wq2
       ...
     job_variables: "{{ prefect.blocks.json-block.value.job-vars }}" # {"image": "cool-image"}
```

### Additional context

I actually wrote already some code which implements the proposed behavior. But I thought it might deviate too much from the fix for #10388. So I wanted ask first, if the feature makes sense. I would also have to implement the same behavior for Variables, but would be happy to do so. ☺️

### PR DESCRIPTION
**Pull request by bjarneschroeder**
Fixes #10388 and closes #10930.

Block references are now being successfully resolved in `prefect.yaml` outside of the `build/push/pull` sections.
As a dict or list parsed content of blocks can now be referenced via a 'keypath'.

Tested with added unit tests and example.

### Example

**Minimal example:**

Content of `work-pools` JSON Block:
```json
{
  "ecs": {
     "name": "ecs-work-pool",
   },
   "ny4": {
     "names": [
       "ny4-first-pool",
       "ny4-second-pool",
    ]
  }
}
```

Value for `default-work-queue-name` String Block:
```text
default-work-queue
```

`prefect.yaml` (slight deviation from #10388 to demonstrate keypath usage):
```yaml
definitions:
    work_pools:
        ecs_work_pool: &ecs_work_pool
            name: "{{ prefect.blocks.json.work-pools.value.ecs.name }}"
            work_queue_name: "{{ prefect.blocks.string.default-work-queue-name }}"
            job_variables:
                image: "*****.dkr.ecr.us-east-2.amazonaws.com/py310-base:0.0.3"

        ny4_work_pool: &ny4_work_pool
            name: "{{ prefect.blocks.json.work-pools.value.ny4.names[1] }}"
            work_queue_name: default
            job_variables:
                image: "****.dkr.ecr.us-east-2.amazonaws.com/py310-base:0.0.3"


# the deployments section allows you to provide configuration for deploying flows
deployments:
- name: test
  version: 0.1
  tags: []
  description: "test flow"
  schedule: {}
  entrypoint: "flow.py:test_flow"
  parameters: {}
  work_pool: *ecs_work_pool
- name: another-test
  version: 0.1
  tags: []
  description: "test flow"
  schedule: {}
  entrypoint: "flow.py:test_flow"
  parameters: {}
  work_pool: *ny4_work_pool
```

`flow.py`:
```python
from prefect import flow

@flow
def test_flow():
    print("I am a very good test")
```

Following these steps you can check out the result:

1. Create a new directory and copy `flow.py` and `prefect.yaml` files in there.
2. Have a Server/Cloud Workspace somewhere running and make sure your prefect config points to it.
3. Create the two blocks with the specified content with the UI.
4. Create the work pools `ecs-work-pool` and `ny4-second-pool`. (Can this be done via the API btw?)
5. Run `prefect deploy --all`.
6. You should see in the UI that everything was resolved correctly or/and run `prefect deployment inspect test-flow/test`/`prefect deployment inspect test-flow/another-test`

**Some additional info about the keypath feature**:

Take the base files from above to try it out yourself.

- If a path cannot be resolved an error is raised:

> ```yaml
> definitions:
>     work_pools:
>         ecs_work_pool: &ecs_work_pool
>             name: "{{ prefect.blocks.json.work-pools.value.ecs.this-key-does-not-exist }}"  # <----
>             work_queue_name: "{{ prefect.blocks.string.default-work-queue-name }}"
>             job_variables:
>                 image: "*****.dkr.ecr.us-east-2.amazonaws.com/py310-base:0.0.3"
>                 
>                 
> #Output
> Invalid template: '{{ prefect.blocks.json.work-pools.value.ecs.this-key-does-not-exist }}'. Found 'value' keyword with additional keys in template, but could not resolve the keypath in the block 
> document data.
> ```

- If a block is referenced whose content is not parsed as a dict or list internally an error is raised:

> ```yaml
> definitions:
>     work_pools:
>         ecs_work_pool: &ecs_work_pool
>             name: "{{ prefect.blocks.string.default-work-queue-name.value.not-a-dict }}"  # <----
>             work_queue_name: "{{ prefect.blocks.string.default-work-queue-name }}"
>             job_variables:
>                 image: "*****.dkr.ecr.us-east-2.amazonaws.com/py310-base:0.0.3"
>                 
> # Output:
> Invalid template: '{{ prefect.blocks.string.default-work-queue-name.value.not-a-dict }}'. Found 'value' keyword with additional keys in reference, but the content of the block was not parsed as a 
> dict or list.
> ```

- You can traverse lists and nested keys. Take a look at `tests/utilities/test_templating:TestResolveBlockDocumentReferences` for that.

Hope you like it, looking forward to your feedback! 😊 

### Checklist

- [x] This pull request references any related issue by including "closes `<link to issue>`"
	- If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect/issues/new/choose) first.
- [x] This pull request includes tests or only affects documentation.
- [x] This pull request includes a label categorizing the change e.g. `fix`, `feature`, `enhancement`, `docs`.
  

For documentation changes:

- [ ] This pull request includes redirect settings in `netlify.toml` for files that are removed or renamed

### CODE REVIEW

**Review by desertaxle — COMMENTED** [review]
Thanks for the PR @bjarneschroeder! 

The implementation is looking good, but I think we can drop the `.value` prior to accessing keys within a block document since block documents are already resolved to their value without `.value`.

**Comment by bjarneschroeder** [comment]
Hey @desertaxle, thank you for taking a look. 

How do you feel about actually keeping the `.value` and therefore being able to also reference different attributes other than `value` in blocks? I just noticed `document.data` always(?) is a dict containing all the attributes. This way all parts of the block are resolvable. The downside might be, that one has to always refer `value` or the specific fields. But it should be consistent that way across blocks.

I implemented the proposed behavior and added another test if you want to check it out. ☺️

**Comment by desertaxle** [comment]
> How do you feel about actually keeping the `.value` and therefore being able to also reference different attributes other than `value` in blocks? I just noticed `document.data` always(?) is a dict containing all the attributes. This way all parts of the block are resolvable. The downside might be, that one has to always refer `value` or the specific fields. But it should be consistent that way across blocks.

We should treat each placeholder like it returns `data` from the block document as a dictionary. This means that blocks like `String` and `DateTime` could have a `.value` after the initial placeholder, but not blocks that don't have a `value` key. Instead, you could access any existing keys for that block document. We'd need to preserve the behavior for a block document with a `value` key like `String` to automatically resolve for backward compatibility.

Apologies for the long delay in responding! Let me know if there's any other way I can help with this PR!

**Comment by bjarneschroeder** [comment]
Hey @desertaxle, I am currently busy at work. But I do have some free time coming up in the following weeks. I will definitely get back to this very soon. ☺️

**Comment by bjarneschroeder** [comment]
I'm on it!

**Comment by bjarneschroeder** [comment]
I think I got it, however when I push to my repo some compatibility tests are failing:

https://github.com/bjarneschroeder/prefect/actions/runs/7273075456/job/19816365478

Is this happening because it is not the "official" repo or due to my changes?

**Comment by desertaxle** [comment]
> I think I got it, however when I push to my repo some compatibility tests are failing:
> 
> https://github.com/bjarneschroeder/prefect/actions/runs/7273075456/job/19816365478
> 
> Is this happening because it is not the "official" repo or due to my changes?

It looks like those compatibility tests pass here, but the unit tests are failing.

**Comment by bjarneschroeder** [comment]
Hi @serinamarie, 
thank you for triggering a re-run and sorry that it did not go through. I was always refreshing the PR via the GUI in GitHub to keep it ready to merge and did not think the tests would fail again. I should have checked better and I'm terribly sorry to waste time and computing resources here!

I adjusted the added test and I think the tests should run now succesfully again. (To my aid, they still did up until the last merge 0af8b9fb25b2a8914f3d3859db66edf12988d609 😄)


As it is obvious with this PR, I struggle to run the tests properly. Running all the tests also takes a lot of time on my machine, but this is what I try to do so. When I run all of the unit tests offline, some arbitrary ones are failing. Same ones go through however, if I execute them for themselves again. Some examples in my case were:

- `tests/cli/test_artifact.py::test_listing_artifacts_when_none_exist`
- `tests/cli/test_block.py::test_listing_blocks_after_saving_a_block`
- `test_kill_passes_other_k8s_api_errors_through`

When I commit to my fork, Docker Build Jobs and most of the integration tests are failing. I saw that you already created an issue for the integration tests. I also get why the docker jobs (login credentials are missing) are not going through. Unit tests on the other hand are not picked up on my fork, so I can't verify there, that my implementation and tests are working really.

**Long story short**: I apologize for all the inconvenience that I create and the time this takes to review. I'm trying my best and I hope you guys can see where I'm coming from. Hopefully we can soon merge something. I'm happy to help! ❤️

**Comment by serinamarie** [comment]
Hi @bjarneschroeder, I did not see all your previous test runs, but at least the last time around, the only test failing was the one you added. In this case, and in future, you can probably just run the tests you add, then the whole file:

I cloned your fork and ran the test locally
```bash
pytest tests/cli/test_deploy.py -k test_deploy_resolves_block_references_in_deployments_section 
```

then I ran the full file locally
```bash
pytest tests/cli/test_deploy.py -vv
```

and I was satisfied to rerun the full test suite :)

**Review by desertaxle — COMMENTED** [review]
This is looking amazing! Thanks so much for working on this @bjarneschroeder! I have a couple of comments on documentation, but the implementation looks great to me!

**Comment by bjarneschroeder** [comment]
> This is looking amazing! Thanks so much for working on this @bjarneschroeder! I have a couple of comments on documentation, but the implementation looks great to me!

The suggestions look way cleaner. I will merge the updates on main and then commit the suggestions from you.

**Review by desertaxle — APPROVED** [review]
LGTM!

**Comment by bjarneschroeder** [comment]
Thank you for your support!🫡😊
